Support lifecycle hooks

yekibud opened the pull request:
<!--
Thank you for contributing to couchdb-helm. Before you submit this PR we'd like to
make sure you are aware of the chart technical requirements and best practices:

* https://github.com/helm/charts/blob/master/CONTRIBUTING.md#technical-requirements
* https://github.com/helm/helm/tree/master/docs/chart_best_practices

For a quick overview across what we will look at reviewing your PR, please read
our review guidelines:

* https://github.com/helm/charts/blob/master/REVIEW_GUIDELINES.md

Following our best practices right from the start will accelerate the review process and
help get your PR merged quicker.

When updates to your PR are requested, please add new commits and do not squash the
history. This will make it easier to identify new changes. The PR will be squashed
anyways when it is merged. Thanks.

Please make sure you test your changes before you push them.
-->

#### What this PR does / why we need it:

Adds support for container lifecycle hooks which serve different use cases than helm hooks and can be more reliable for similar use cases (e.g. for changes outside of helm management like a pod or statefulset being deleted).

#### Special notes for your reviewer:

My use case requires a sophisticated postStart hook with variable interpolation from a parent chart, but I figured many users of this chart may not use it as a subchart, so I included the option to pass in simple container hooks via values, as well.  I tried to make this opaque to the user so they could just use `.Values.lifecycle` in both cases, but I couldn't figure out how to get helm to differentiate between a template and a values map since both are `map` types.  So I wound up supporting both use cases with `.Values.lifecycle` and `.Values.lifecycleTemplate`.

Furthermore, following the same pattern, I also added `.Values.extraEnv` and `.Values.extraEnvTemplate` to support passing secrets to the hooks.

#### Checklist
- [x] Chart Version bumped
- [x] e2e tests pass
- [x] Variables are documented in the README.md

I get the following error when trying to run `make test`.  Not sure if this is an existing issue, but it looks unrelated to the changes in this PR.

```
Error: failed linting and installing charts: failed identifying merge base: must be in a git repository
------------------------------------------------------------------------------------------------------------------------
No chart changes detected.
------------------------------------------------------------------------------------------------------------------------
failed linting and installing charts: failed identifying merge base: must be in a git repository
Removing ct container...
Deleting cluster "chart-testing" ...
Done!
make: *** [Makefile:31: test] Error 1
```

yekibud commented:
@willholley any feedback?  In case the use case of the PR isn't clear, here's a more concrete example:

Let's say you need to iterate through a list of values to create admin users, or some other dynamic process after your couchdb instance is stood up.  This PR enables you to do that by providing a template to the statefulset that will allow you to populate those values.

Sound useful?

willholley commented:
Thanks for the PR @yekibud. I don't see any issues with this so can merge once it's rebased, tests pass etc.

yekibud commented:
> Thanks for the PR @yekibud. I don't see any issues with this so can merge once it's rebased, tests pass etc.

@willholley Thanks.  Rebased and updated my branch.  I get the same error I reported for `make test` on `main` so I don't think that has to do with my PR.

yekibud commented:
@willholley sorry for the delay.  Resolved the latest merge 

> > Thanks for the PR @yekibud. I don't see any issues with this so can merge once it's rebased, tests pass etc.
> 
> @willholley Thanks. Rebased and updated my branch. I get the same error I reported for `make test` on `main` so I don't think that has to do with my PR.

@willholley resolved the latest merge conflicts, but tests still don't run for me on my branch or on main, per my message in the PR summary.  Thoughts?

big-r81 commented on couchdb/Chart.yaml:
```diff
@@ -18,4 +18,4 @@ maintainers:
   - name: willholley
     email: willholley@apache.org
-icon: http://couchdb.apache.org/CouchDB-visual-identity/logo/CouchDB-couch-symbol.svg
+icon: http://couchdb.apache.org/CouchDB-visual-identity/logo/CouchDB-couch-symbol.svg
```
small nit: missing nl

yekibud replied:
Ohhh, the linter _wants_ a newline at the end of the file.  Okay, that's new to me.  Fixed - although, looks like the linter chose to ignore the previous linting error in the values.yaml, which didn't appear to have a newline, and complained, but after I edited the file in the Github editor, it didn't complain, although there still isn't a newline.   In any case, good now, I think?

yekibud commented:
> @yekibud personally, I would squash the commits into one commit and then pass the baton to @willholley :-)

Heh, I think the instructions for this repo say _not_ to squash, right?

big-r81 commented:
> > @yekibud personally, I would squash the commits into one commit and then pass the baton to @willholley :-)
> 
> Heh, I think the instructions for this repo say _not_ to squash, right?

Does it? Nevermind, Will already does the job ;-)
